api/base/middleware.py

```diff
@@ -1,70 +1,72 @@
-from pymongo.errors import OperationFailure
 from raven.contrib.django.raven_compat.models import sentry_exception_handler
 
+from framework.mongo.handlers import (
+    connection_before_request,
+    connection_teardown_request
+)
 from framework.tasks.handlers import (
     celery_before_request,
+    celery_after_request,
     celery_teardown_request
 )
-from framework.transactions import commands, messages, utils
+from framework.transactions.handlers import (
+    transaction_before_request,
+    transaction_after_request,
+    transaction_teardown_request
+)
 
 from .api_globals import api_globals
 
-from api.base import settings
 
+class MongoConnectionMiddleware(object):
+    """MongoDB Connection middleware."""
+
+    def process_request(self, request):
+        connection_before_request()
+
+    def process_response(self, request, response):
+        connection_teardown_request()
+        return response
 
-# TODO: Verify that a transaction is being created for every
-# individual request.
-class TokuTransactionsMiddleware(object):
-    """TokuMX transaction middleware."""
+class CeleryTaskMiddleware(object):
+    """Celery Task middleware."""
+
+    def process_request(self, request):
+        celery_before_request()
+
+    def process_exception(self, request, exception):
+        """If an exception occurs, clear the celery task queue so process_response has nothing."""
+        sentry_exception_handler(request=request)
+        celery_teardown_request(error=True)
+        return None
+
+    def process_response(self, request, response):
+        """Clear the celery task queue if the response status code is 500 or above"""
+        celery_after_request(response, base_status_code_error=500)
```

Comment:
The jobs should die, as for the 400 vs. 500.

A. We didn't clear them at all before.
B. I'm still not certain there isn't  case when a valid task (email?) might be sent w/ a 400'ish response... thoughts?